### DISCUSSION TITLE
리팩터링 4주차 안건

### DISCUSSION THREAD

**Discussion by punchdrunkard** [prgrms-web-devcourse/Team-BackFro-ArtZip-FE#276] [02.Refactoring]
.

**Comment by punchdrunkard** [comment]
메모) `useSWR`의 캐시 문제 (윤정님이 언급해주신 이슈)

**Reply by gitul0515** [reply]
swr을 통해 방법을 찾아보고, 안 되면 useEffect로 돌리기

**Reply by punchdrunkard** [reply]
## 기홍님 슬랙 답변

@gitul0515 

다은님, 커뮤니티 페이지에서 좋아요 버그의 원인을 찾은 것 같습니다. :bug:

설명해보겠습니다.
이미 알고 계시겠지만,
**SWR**은 캐시된 데이터를 먼저 가져온 뒤,
서버에서 데이터의 업데이트를 확인하고,
업데이트가 발생했다면 최신 데이터를 가져옵니다.
로그를 찍어보면 이것은 정상적으로 이루어지고 있습니다.

첫 번째 **ReviewFeed**에 좋아요를 누른 뒤,
페이지를 나갔다가 다시 들어왔다고 가정해봅시다.

![image](https://user-images.githubusercontent.com/74234333/191030187-f3d163fd-160b-4956-9b15-87545f1b2ef5.png)

**빨간색**은 캐시된 데이터이고, **파란색**은 최신 데이터입니다.

![image](https://user-images.githubusercontent.com/74234333/191030368-f14dfc8b-cd2b-48b6-a2ee-80d8408c6caf.png)

[9:36](https://greppkdt-webdev.slack.com/archives/C03NYDFRMQW/p1663590995350559)
빨간색에서는 isLiked가 false입니다. 예전 데이터입니다.
파란색에서는 isLiked가 true입니다. 최신 데이터입니다.

![image](https://user-images.githubusercontent.com/74234333/191030426-844afe65-ec11-409a-9fd5-565c70bbb52c.png)
![image](https://user-images.githubusercontent.com/74234333/191030544-c16a5c2b-648d-456a-9d5b-27a4ffb054b7.png)

그렇다면 왜 해당 **ReviewFeed**에는 반영되지 않았을까요?
추측하건대, props로 전달되는 **feed가 객체**이기 때문입니다.
객체의 속성(isLiked, likeCount)는 변경되었으나,
객체의 참조값은 변경되지 않았으므로,
해당 ReviewFeed는 리렌더링이 되지 않은 것으로 추측합니다.

혹은 reviewId가 달라지지 않았기 때문일 수도 있겠군요.
map을 사용해서 key={reviewId}를 넣어주고 있는데
reviewId는 그대로이므로, 리렌더링이 발생하지 않았다는 생각도 드네요.
```typescript
// CommunityPage.tsx
          {feeds.map((feed) => {
            const { reviewId, user } = feed;
            return (
              <ReviewFeed
                key={reviewId}
                feed={feed}
                isMyFeed={userId === user.userId}
                onDeleteButtonClick={handleDeleteButtonClick}
              />
            );
          })}
```
![image](https://user-images.githubusercontent.com/74234333/191030687-efac8e56-b944-427e-8478-390e2db668f2.png)

해결 방안은 아래와 같습니다.

![image](https://user-images.githubusercontent.com/74234333/191030768-ad16bac0-f7bf-47c0-8d0c-449f1e186d02.png)

`useEffect`를 사용하여 `isLiked`, `likeCount`의 변화를 관찰하여
최신화된 데이터를 직접 반영해주면 됩니다.
그러면 뒤로가기 후, 다시 들어올 때
데이터의 변경 사항이 제대로 반영되는 것을 확인했습니다.

새로고침 시에도, 물론 반영됩니다.

![image](https://user-images.githubusercontent.com/74234333/191030890-ad050878-c710-40fe-bc99-257fc28b8b31.png)

**Comment by gitul0515** [comment]
맞춤 전시회에 처음 진입할 때에는 10개의 카드가 보이고, 
페이지네이션을 넘기면 8개가 되는 것 같습니다. 

![image](https://user-images.githubusercontent.com/80658269/190563719-3609c20e-098f-4083-af45-4f1ba45025de.png)

![image](https://user-images.githubusercontent.com/80658269/190563733-32fa6d6c-a79e-45ff-9d98-5162326fc5c6.png)

**Reply by gitul0515** [reply]
8개로 통일하기

**Reply by gitul0515** [reply]
그리드 적용 (나중에)

**Comment by gitul0515** [comment]
로그인, 회원가입 버튼에 디바운스를 적용해도 될까요? 
버튼 연타로 발생하는 문제를 방지하려고 합니다. 

![image](https://user-images.githubusercontent.com/80658269/190564229-f74d9640-8bb1-40ba-81d5-4b7c2a518158.png)

**Reply by gitul0515** [reply]
디바운스 적용 땅땅

**Comment by gitul0515** [comment]
커뮤니티 페이지에서 무한 스크롤이 동작하지 않고 있는데, 
이것이 SWR로 인한 문제인가요?

**Reply by punchdrunkard** [reply]
앗 제 쪽에서는 정상적으로 작동하네요...?

**Reply by gitul0515** [reply]
현재 배포에서는 첫 20개만 나오는 것 같아요 :)

**Reply by gitul0515** [reply]
intersectionObserver로 수정 예정. 다시 확인해주기

**Comment by gitul0515** [comment]
리팩토링 회의 마지막 날짜

**Reply by gitul0515** [reply]
계속~

**Reply by gitul0515** [reply]
ㅎㅎ

**Comment by gitul0515** [comment]
스켈레톤 추가 여부 (우선순위 낮음)

**Reply by gitul0515** [reply]
짬이 되면 적용해보기

**Reply by punchdrunkard** [reply]
Optimize Cumulative Layout Shift (CLS)

**Comment by gitul0515** [comment]
배포 버전 추가

**Reply by gitul0515** [reply]
dev.artzip.shop를 등록해보고, 백엔드분께 CORS 허용을 요청

**Comment by gitul0515** [comment]
컴포넌트 테스트

**Comment by gitul0515** [comment]
develop 브랜치를 main 브랜치에 머지 후, 윤정님께 고지
release 업데이트
